### PR TITLE
[ASTextNode2] Upgrade lock safety by protecting all ivars (including rarely-changed ones).

### PR DESCRIPTION
**Pull request by appleguy**
Although I don't know of any specific crashes caused by this, I think we should
lock all properties by default. There are also some indications of premature
optimization in keeping lock scope small, where it is actually important to
have transactional integrity, and also where the ASDisplayNode base class is
otherwise going to repeatedly re-lock the object anyway.

I think this will remain pretty efficient, especially with os_unfair_lock enabled.

### CODE REVIEW

File: Source/ASTextNode2.mm [hunk 1]
```diff
@@ -910,16 +927,14 @@ - (CGFloat)shadowOpacity
...
+  ASLockScopeSelf();
+
+  if (_shadowOpacity == shadowOpacity) {
+    return;
   }
-  
+
+  _shadowOpacity = shadowOpacity;
+
   [self setNeedsDisplay];
```
Comment (Adlai-Holler) [inline Source/ASTextNode2.mm]: Great place to use `ASLockedSelfCompareAssign`!

File: Source/ASTextNode2.mm [hunk 2]
```diff
@@ -986,29 +998,36 @@ - (NSArray *)pointSizeScaleFactors
...
   }
 }
 
 - (void)setTruncationAttributedText:(NSAttributedString *)truncationAttributedText
 {
-  if (ASLockedSelfCompareAssignCopy(_truncationAttributedText, truncationAttributedText)) {
+  ASLockScopeSelf();
+  if (!ASObjectIsEqual(_truncationAttributedText, truncationAttributedText)) {
+    _truncationAttributedText = [truncationAttributedText copy];
     [self _invalidateTruncationText];
```
Comment (Adlai-Holler) [inline Source/ASTextNode2.mm]: Why this change? Is it so that we now call `_invalidateTruncationText` with the lock held?

File: Source/ASTextNode2.mm [hunk 3]
```diff
@@ -210,6 +212,7 @@ - (BOOL)supportsLayerBacking
 
 - (void)setTextContainerInset:(UIEdgeInsets)textContainerInset
 {
+  ASLockScopeSelf();
   BOOL needsUpdate = !UIEdgeInsetsEqualToEdgeInsets(_textContainer.insets, textContainerInset);
   _textContainer.insets = textContainerInset;
```
Comment (Adlai-Holler) [inline Source/ASTextNode2.mm]: Good place for `ASLockedSelfCompareAssignCustom`

File: Source/ASTextNode2.mm [hunk 4]
```diff
@@ -189,9 +190,10 @@ - (BOOL)supportsLayerBacking
   if (!super.supportsLayerBacking) {
     return NO;
   }
-  
+
+  ASLockScopeSelf();
   // If the text contains any links, return NO.
-  NSAttributedString *attributedText = self.attributedText;
+  NSAttributedString *attributedText = _attributedText;
```
Comment (Adlai-Holler) [inline Source/ASTextNode2.mm]: Since we just took the lock, can we go back to using the ivar here?

File: Source/ASTextNode2.mm (hunk 2 above)
Comment (appleguy) [reply Source/ASTextNode2.mm]: Yes, otherwise there is no transactional integrity.

**Comment by Adlai-Holler** [comment]
Just to note: `_textContainer` is a true invariant of the node and never changes. Its properties are atomic. We should hold the lock when making batch accesses to properties but otherwise it can be accessed without the lock. Its `-copy` method is also atomic.

File: Source/ASTextNode2.mm (hunk 2 above)
Comment (Adlai-Holler) [reply Source/ASTextNode2.mm]: OK in cases such as this (side effect keeps lock), we can use the non-locked compare-assign macros, like so:

```
ASLockScopeSelf();
if (ASCompareAssignCopy(_truncationAttributedText, truncationATtributedText)) {
  [self _invalidateTruncationText];
}
```

I set it up to work either way!

Comment (appleguy) [reply Source/ASTextNode2.mm]: Great! This is definitely an improvement.

I'm having trouble getting the minutes to circle back to this; will try, maybe this weekend.
